Port pcl_passthrough_filter to ROS2

Tags01 opened the pull request:
# Changes Made
## passthrough_filter_service:
- Introduced class format, moved main functionality to the constructor, and marked data private.
- Added missing header files.
- Added error checking for all input parameters: all parameters must produce a volumn with a
  positive, non-zero, non-NaN size. Node will shutdown on error. This was moved to its own
  function.
- Convert `fromROSMsg` -> `moveFromROSMsg`.
- Preserve TF frame information in the result.
- replace explicit `std::shared_ptr(new TYPE) ` with definition with `std::make_shared<TYPE>()` format.
- Convert parameter namespace to `std::string_view` to follow google style guidelines.

## simple_test_passthrough_filter:
 - Omit explicit test checks. This script is harder to test accurately since its very sensitive to input data. However, it's easy to verify the code works visually which is included in the script. These tests will be replaced anyways with dedicated unit tests so these can be omitted for now.

## Setup
- Works on `g++ 11.4.0`, `clang++ 19`, and `icpx 2024.2.1`.
- Lints Properly.
- No Valgrind and sanitizer checks yet.

## Misc
- `CMakeLists.txt`: Added notes about why the program must not use `-Ofast` or `-ffast-math`
- `CMakeLists.txt`: Added the new files so they can be built
- Added two new launch files. These are mostly the same as with other commits. The default parameters for the bounding box in passthrough_filter default to a `1x1x1 m` cube centered at `(0, 0, 0)` 

# Evidence of Working Code
The darker background is the entire field of view of the camera, while the bright color segments only contain the filtered points. The default settings are used.
![ezgif-7-fc5fa396ea](https://github.com/user-attachments/assets/e4a3cabe-484f-4bb2-95f0-a48301307064)
When the arrow is moved up, it falls within the bound of the bottom of the bounding box so the colors appear brighter on the image.

Tags01 committed:
NOT BUILT AND NOT TESTED; See Notes
- remove redundant smart pointer
- put initial implementation in a class
- use `moveFromROSMsg` instead of `fromROSMsg`
- update includes

Tags01 committed:
Merge branch 'uml-robotics:main' into ct_passthrough_filter

Tags01 committed:
Builds, Lints, Visual test passes. See notes:
Converted pcl_passthrough_filter from ROS1 -> ROS2
Tested in Intel 2024, Clang++-19, G++ 4.8

passthrough_filter:
- introduced class format, moved main functionality to the constructor, and marked data private.
- added missing header files
- added error checking for all input parameters: all parameters must produce a volumn with a
  positive, non-zero, non-NaN size. Node will shutdown on error. This was moved to its own
  function.
- Convert `fromROSMsg` -> `moveFromROSMsg`.
- preserve TF frame information in the result.
- explicit shared ptr definition with `std::make_shared`.
- convert parameter namespace to `std::string_view` to follow google style guidelines.

simple_test:
 - omit testing functionality, this is harder to test but is easy to verify visually.
   These tests will be replaced anyways.

CMake:
 - Added executables
 - Added notes about why the program must not use `-Ofast` or `-ffast-math`

Added appropriate launch files for the node and testing (includes comments)

Tags01 committed:
Builds, tests pass. Removed a word from a comment.

GregoryLeMasurier commented on pcl_utility_msgs/srv/PCLPassthroughFilter.srv:
This is the same as the voxel grid service message - it helps make it easier to read this way (and it's the same way it was done before) but I wonder if we can reduce duplicated messages some way. Let's talk about this with the group next week

GregoryLeMasurier commented on pcl_utilities/src/passthrough_filter_service.cpp:
```diff
@@ -0,0 +1,186 @@
+    pass_x.setInputCloud(temp_cloud);
+    pass_x.setFilterFieldName("x");
+    pass_x.setFilterLimits(x_min_, x_max_);
+    // pass_x.setFilterLimitsNegative(false);
```
Do we need the commented out line?

GregoryLeMasurier reviewed (changes requested):
Awesome! One small thing to clean up

Tags01 committed:
Builds, Lints, Tests Pass. No sanitizer checks.
Checked in Clang++ and g++

Changes:
- Added explicit dependencies to CMakeLists.txt/package.xml (no longer transitive)
- Removed std::move on lvalue reference argument
- Added explicit overflow checks and eliminating warnings in clang about casts.
     Note: there are still warning about `-Wdouble-promotion`. This diagnostic is only
     supposed to trigger with implicit casts. However there is a diagnostic issue where
     explicit braced casts are also included.

Tags01 commented:
# Changes
- Added explicit dependencies for `rcl_interfaces`/`rclcpp`. These dependencies were missed because they are always included transitively. Dependencies in the CMakeLists.txt/package.xml are now alphabetical.
- Added explicit overflow checks when converting from double to float. This also resolves warnings in Clang with `-Wimplicit-float-conversions`
- Added comments to the launch script
- Changed out lvalue parameter to rvalue parameter in the test. Beforehand I used `std::move` on an lvalue parameter.

Tags01 committed:
Added Dates Changes

GregoryLeMasurier commented on pcl_utilities/src/tests/simple_test_passthrough_filter.cpp:
```diff
@@ -0,0 +1,125 @@
+    }
+  }
+
+  void process_point_cloud(PointCloud2 && point_cloud)
```
Why did we change to pass in an address of an address? Is this best practice? It seems to add overhead where we have to use std::move before each call, where it was handled for the user before

GregoryLeMasurier replied:
Addressed in person

GregoryLeMasurier commented on pcl_utilities/src/passthrough_filter_service.cpp:
```diff
@@ -0,0 +1,203 @@
+private:
+  rclcpp::Service<PCLPassthroughFilter>::SharedPtr passthrough_filter_service_;
+  rcl_interfaces::msg::ParameterDescriptor constraints_;
+  double x_min_;
```
Why did we switch from float to double and then add in a bunch of checks to make sure the conversion happened correctly? Why can't we just use floats?

GregoryLeMasurier commented on pcl_utilities/src/tests/simple_test_passthrough_filter.cpp:
```diff
@@ -0,0 +1,125 @@
+ *    `ros2 launch pcl_utilities test_passthrough_filter_service.xml point_cloud_topic:=<POINT_CLOUD_TOPIC>`
+ */
+#include <chrono>   // std::chrono::seconds
+#include <cstddef>  // size_t
```
Not used in file

GregoryLeMasurier commented on pcl_utilities/src/tests/simple_test_passthrough_filter.cpp:
```diff
@@ -0,0 +1,125 @@
+ */
+#include <chrono>   // std::chrono::seconds
+#include <cstddef>  // size_t
+#include <cstdint>  // int64_t
```
Not used in this file

GregoryLeMasurier commented on pcl_utilities/src/tests/simple_test_passthrough_filter.cpp:
```diff
@@ -0,0 +1,125 @@
+#include <chrono>   // std::chrono::seconds
+#include <cstddef>  // size_t
+#include <cstdint>  // int64_t
+#include <functional>  // std::bind, std::placeholders
```
Not used in this file

GregoryLeMasurier commented on pcl_utilities/src/tests/simple_test_passthrough_filter.cpp:
```diff
@@ -0,0 +1,125 @@
+#include <cstddef>  // size_t
+#include <cstdint>  // int64_t
+#include <functional>  // std::bind, std::placeholders
+#include <ios>  // std::fixed, std::setprecision
```
Not used in this file

GregoryLeMasurier commented on pcl_utilities/src/tests/simple_test_passthrough_filter.cpp:
```diff
@@ -0,0 +1,125 @@
+#include <functional>  // std::bind, std::placeholders
+#include <ios>  // std::fixed, std::setprecision
+#include <memory>  // std::make_shared
+#include <sstream>  // std::stringstream
```
Not used in this file

GregoryLeMasurier reviewed (changes requested):
Forgot to check includes - small cleanup

Tags01 committed:
Builds, Lints, Visual Tests Pass. Removed redundant includes.

flynnbm reviewed (approved):
Modifications to the original and overall format looks great
